### PR TITLE
Monitor abilities: register jetpack-monitor category via Registrar

### PR DESCRIPTION
**Pull request by enejb**
Fixes #

## Proposed changes

* Register two WordPress Abilities API abilities for Jetpack Downtime Monitor via the shared `Registrar` base class:
  * `jetpack-monitor/get-monitor-status` — zero-arg read returning `{ module_active, user_connected, notifications_enabled, last_status_change }` on the happy path. `last_status_change` is the timestamp of the most recent up/down transition (the legacy `last_status_change` projection), or `null` when no transition has been recorded yet — the only legitimate null on the success path. Fails with `WP_Error` codes when preconditions aren't met: `jetpack_monitor_module_inactive` (defensive — unreachable in practice), `jetpack_monitor_not_connected` (with a steering message pointing the caller to the My Jetpack admin page), or `jetpack_monitor_service_unreachable` when either underlying XML-RPC read fails. This replaces the earlier null-fields-on-failure behavior so agents get an actionable next step.
  * `jetpack-monitor/set-notifications` — declarative state-setter taking `{ enabled }`, idempotent; returns `{ enabled, changed }`. Errors with `jetpack_monitor_not_connected` when the user isn't connected to Jetpack; `jetpack_monitor_module_inactive` is reachable only via a race (see below). Either outcome (changed write **or** no-op) syncs the local `monitor_receive_notifications` option to the known-good remote value, so the legacy `Jetpack_Core_Json_Api_Endpoints::get_remote_value` reader — which trusts that option before falling back to a remote read — can never surface a stale state.
* Both abilities set `show_in_rest` **and** `meta.mcp` (`public => true`, `type => tool`), matching the sibling `jetpack/get-modules` / `jetpack/set-module-status` definitions, so they're discoverable by MCP/agent tooling rather than REST-only.
* Wire `Monitor_Abilities::init()` from `modules/monitor.php`, which Jetpack only loads when the Monitor module is active. **The Monitor abilities are therefore only registered while the Monitor module is active** — `wp_get_abilities()` will not return `jetpack-monitor/*` slugs when Monitor is inactive. This keeps the always-on ability surface minimal; future module-management abilities (out of scope here) will be the discovery/activation entry point for module-gated categories like this one.
* `Registrar::init()` handles the `jetpack_wp_abilities_enabled` gate, category/abilities lifecycle hooks, and WP < 6.9 compatibility — no manual `add_action` plumbing required.
* Added as a minor changelog entry under `enhancement`.

## Related product discussion/links

* Tracking surface for Radical Speed Month — "Abilities Everywhere".

## Does this pull request change what data or activity we track or use?

No. The abilities wrap the existing `jetpack.monitor.isUserInNotifications`, `jetpack.monitor.setNotifications`, and `jetpack.monitor.getLastDowntime` XML-RPC endpoints that the Monitor module already uses. No new data is collected or transmitted.

## Testing instructions

1. Install/activate Jetpack from this branch and connect the site + user.
2. Opt in to the Abilities API:
   ```php
   add_filter( 'jetpack_wp_abilities_enabled', '__return_true' );
   ```
   (mu-plugin or snippet)
3. With the Monitor module **active**, confirm the abilities register:
   ```bash
   wp eval 'foreach ( wp_get_abilities() as $a ) { if ( str_starts_with( $a->get_name(), "jetpack-monitor/" ) ) echo $a->get_name() . "\n"; }'
   ```
   Expected output:
   ```
   jetpack-monitor/get-monitor-status
   jetpack-monitor/set-notifications
   ```
4. REST surface check:
   ```bash
   curl -s -u admin:app_password https://example.test/wp-json/wp-abilities/v1/abilities | jq '[.[] | select(.name | startswith("jetpack-monitor/"))] | length'
   ```
   Expect `2`.
5. MCP metadata check — confirm both abilities advertise the public MCP tool descriptor:
   ```bash
   wp eval 'foreach ( array( "jetpack-monitor/get-monitor-status", "jetpack-monitor/set-notifications" ) as $s ) { var_export( wp_get_ability( $s )->get_meta()["mcp"] ?? null ); echo "\n"; }'
   ```
   Each should print `array ( 'public' => true, 'type' => 'tool' )`.
6. With the Monitor module **inactive** (`wp jetpack module deactivate monitor`), re-run the step 3 command — expect **empty output**. The abilities load alongside the module by design.
7. With the Monitor module **active** and the user connected, exercise `get-monitor-status`:
   ```bash
   wp eval 'var_export( wp_get_ability( "jetpack-monitor/get-monitor-status" )->execute( [] ) );'
   ```
   Expect an array with `module_active=true`, `user_connected=true`, a boolean `notifications_enabled`, and either a `"YYYY-MM-DD HH:mm:ss"` UTC string or `null` for `last_status_change` (the "no transition recorded yet" signal).
8. Exercise idempotency on `set-notifications`, then confirm the local option mirrors remote even on a no-op:
   ```bash
   wp eval 'var_export( wp_get_ability( "jetpack-monitor/set-notifications" )->execute( [ "enabled" => true ] ) );'
   wp eval 'var_export( wp_get_ability( "jetpack-monitor/set-notifications" )->execute( [ "enabled" => true ] ) );'
   wp option get monitor_receive_notifications
   ```
   First call → `[ 'enabled' => true, 'changed' => true ]`; second call → `[ 'enabled' => true, 'changed' => false ]`. The `monitor_receive_notifications` option should read `1` after the no-op (mirrored from the remote value), even if it was previously stale.
9. Run with a disconnected user — both abilities should now surface an actionable `WP_Error` instead of silent nulls:
   ```bash
   wp eval 'var_export( wp_get_ability( "jetpack-monitor/get-monitor-status" )->execute( [] ) );'
   wp eval 'var_export( wp_get_ability( "jetpack-monitor/set-notifications" )->execute( [ "enabled" => true ] ) );'
   ```
   Both should return a `WP_Error` with code `jetpack_monitor_not_connected`. The `get-monitor-status` error message should mention "My Jetpack" — that's the steering hint for callers.
10. PHPUnit:
    ```bash
    cd projects/plugins/jetpack && composer phpunit -- --filter Monitor_Abilities_Test
    ```
    Covers Registrar wiring, abstract getters, permissions, schema edge cases, success-path coverage (`changed:true` / `changed:false`, including the no-op local-option sync) for `set-notifications`, the public MCP tool metadata on both abilities, the new `WP_Error` paths on `get-monitor-status` (module inactive, user not connected, both `service_unreachable` variants), the `normalize_last_status_change` normalizer (empty/whitespace/zero-date/valid/non-string), and the negative bootstrap test confirming abilities are NOT registered when the Monitor module is inactive.

Permissions: both abilities are gated on `current_user_can( 'jetpack_admin_page' )`, matching the legacy Monitor admin UI. Notifications are a per-user preference, so the same gate is appropriate for the write.

### CODE REVIEW

File: projects/plugins/jetpack/modules/monitor.php [hunk 1]
```diff
@@ -138,3 +138,6 @@ public function monitor_get_last_downtime() {
 }
 
 new Jetpack_Monitor();
+
+// Register Jetpack Monitor abilities (WordPress Abilities API, WP 6.9+).
+\Automattic\Jetpack\Plugin\Abilities\Monitor_Abilities::init();
```
[the remark this answers is not in the document]
Comment (enejb) [reply projects/plugins/jetpack/modules/monitor.php:143]: Thanks — keeping the gated registration here intentionally:

- The design is "module-specific abilities load alongside the module," matching how Jetpack lazy-loads other module-only code.
- Discovery of the inactive state isn't lost: an AI client looking to configure Monitor would call `jetpack-modules/set-module-status` with `slug: "monitor"` and `active: true` first; `set-notifications`'s description already routes the agent that way (`Fails with jetpack_monitor_module_inactive (call jetpack-modules/set-module-status …)`).
- This keeps the always-on ability surface minimal, which we'd like to keep small as more module-specific abilities land.

So `get-monitor-status` is a *post-activation* diagnostic, not a discovery tool — that role belongs to the modules-management abilities.

[the remark this answers is not in the document]
Comment (chrisbliss18) [reply projects/plugins/jetpack/modules/monitor.php:143]: I had a similar concern to Copilot. I see what you're saying, but the PR describes and tests a different contract. Your PR description and testing instructions say to call `get-monitor-status` while Monitor is inactive and expect `module_active: false`. Unless I'm missing something, that cannot work if the ability is not registered while Monitor is inactive. Also, the recovery path depends on `jetpack-modules/set-module-status`, but I could not find that ability in this branch.

Am I missing something or is your explanation here and the PR text not syncing up?

File: projects/plugins/jetpack/src/abilities/class-monitor-abilities.php [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+						'destructive' => false,
+						'idempotent'  => true,
+					),
+					'show_in_rest' => true,
+				),
+			),
+
+			'jetpack-monitor/set-notifications'  => array(
+				'label'               => __( 'Set Jetpack Monitor notifications', 'jetpack' ),
+				'description'         => __( 'Enable or disable downtime email notifications for the current user. Idempotent — setting the state to the current value returns changed=false. Returns { enabled, changed }. Preconditions: the Monitor module must be active and the current user must be connected to Jetpack; call jetpack-monitor/get-monitor-status first to verify. Fails with jetpack_monitor_module_inactive (call jetpack-modules/set-module-status with slug "monitor" and active true) or jetpack_monitor_not_connected when preconditions are not met.', 'jetpack' ),
```
Comment (chrisbliss18) [inline projects/plugins/jetpack/src/abilities/class-monitor-abilities.php]: This points agents to `jetpack-modules/set-module-status`, but I couldn't find that ability registered anywhere in this branch.

File: projects/plugins/jetpack/src/abilities/class-monitor-abilities.php [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+					'additionalProperties' => false,
+				),
+				'output_schema'       => array(
+					'type'       => 'object',
+					'properties' => array(
+						'module_active'         => array( 'type' => 'boolean' ),
+						'user_connected'        => array( 'type' => 'boolean' ),
+						'notifications_enabled' => array( 'type' => array( 'boolean', 'null' ) ),
+						'last_downtime'         => array( 'type' => array( 'string', 'null' ) ),
+					),
```
Comment (chrisbliss18) [inline projects/plugins/jetpack/src/abilities/class-monitor-abilities.php:81]: I think `last_downtime` is a little too specific for what this value actually represents.

The WPCOM XML-RPC method ultimately returns `JP_Monitor::get_last_downtime()`, but that reads the legacy `last_status_change` field. That field is maintained for status transitions, not strictly "the last time downtime started." Could this either be renamed to `last_status_change` or have the description updated to make the legacy semantics clearer?

Note: This also matters for Jetmon v2 since the rollout depends on preserving the legacy `last_status_change` projection. Jetmon v2 has event history, so assuming that I get to roll it out, a `last_downtime` could accurately be calculated (given enough time for the data to populate that is).

File: projects/plugins/jetpack/tests/php/src/Monitor_Abilities_Test.php [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+			$name = $ability->get_name();
+			if ( str_starts_with( $name, 'jetpack-monitor/' ) ) {
+				$registered_slugs[] = $name;
+			}
+		}
+
+		foreach ( array_keys( Monitor_Abilities::get_abilities() ) as $slug ) {
+			$this->assertContains( $slug, $registered_slugs, "Ability {$slug} should be registered." );
+		}
+	}
```
Comment (chrisbliss18) [inline projects/plugins/jetpack/tests/php/src/Monitor_Abilities_Test.php:271]: These tests verify the registrar class when called directly, but they don't cover the actual bootstrap location added in `modules/monitor.php`.

That misses the main behavior risk in this PR: when Monitor is inactive, the module file isn't loaded, so the registrar init never runs and the abilities aren't registered. I think there should be a test that exercises the intended inactive-module behavior through the real loading path, or the test expectations should be changed if the abilities are intentionally active-module-only.

**Review by chrisbliss18 — CHANGES_REQUESTED** [review]
Thanks for the work on this. It's nice to see someone interested in Monitor. 😄

I like the overall direction of exposing Monitor through the Abilities API, especially that the implementation reuses the existing XML-RPC methods instead of inventing a new data path. The default-off `jetpack_wp_abilities_enabled` gate also seems like the right shape for a cautious rollout.

I do think there are a few issues that need tightening before this ships which I detailed in code-specific comments.

Recommendations:
1. Add a test that verifies registration behavior with Monitor inactive using the real bootstrap path.
2. Add success-path coverage for `set-notifications`, including the `changed: true` and `changed: false` cases.
3. Update the `last_downtime` name/description to reflect that the WPCOM value is the legacy `last_status_change`.
4. Make the testing instructions include exact ability execution commands or REST calls, not just "run get-monitor-status", so reviewers can reproduce the behavior consistently.

**Comment by enejb** [comment]
Thanks for the thorough review, @chrisbliss18 — you were right that the PR text and the implementation weren't telling the same story. Pushed `ddcd86e905` to align everything around **gated registration** (Monitor abilities only register while the Monitor module is active). Walking through your numbered recommendations:

**Top-level concern (registration location vs. the PR description / `jetpack-modules/set-module-status`).** You called it correctly — the PR text described an always-on contract that the gated registration in `modules/monitor.php` cannot deliver, and the recovery path referenced a `jetpack-modules/set-module-status` ability that doesn't exist on this branch. Rather than expand the always-on surface, I aligned the contract with the implementation:
- Updated the PR description (test step 5 now confirms the abilities are **absent** when Monitor is inactive, since that's the design).
- Removed the `jetpack-modules/set-module-status` reference from `set-notifications`'s description and from the `jetpack_monitor_module_inactive` error message.

**1. Test the real bootstrap path.** Added a negative test `test_abilities_are_not_registered_when_monitor_module_is_inactive` that does NOT call `init()` or fire the lifecycle, and asserts `wp_get_abilities()` returns no `jetpack-monitor/*` slugs. That mirrors the actual production path (where `modules/monitor.php` is only `require`d when the module is active) and codifies the gated-registration contract.

**2. Success-path coverage for `set-notifications`.** Added two tests that drive the callback end-to-end through a `Monitor_Abilities_Test_Stub` subclass:
- `test_set_notifications_returns_changed_true_when_state_flips` — verifies the IXR apply is called exactly once and the `monitor_receive_notifications` option mirrors the new state.
- `test_set_notifications_returns_changed_false_when_state_matches` — verifies the idempotent no-op: apply is **not** called and the mirrored option is **not** written.

To support this without standing up a Jetpack token fixture or hitting a live IXR endpoint, I extracted three `protected static` seams in `Monitor_Abilities` (`is_user_connected_to_jetpack`, `apply_notifications_update`, and made `fetch_notifications_state` overridable). Internal calls switched from `self::` to `static::` for late static binding. No test-only branches in production code — just an extensibility improvement.

**3. `last_downtime` semantics.** Good catch on the legacy `last_status_change` projection — renamed the field accordingly in the output schema, the local variable, and the private fetch method (`fetch_last_downtime` → `fetch_last_status_change`). The transient key stays `monitor_last_downtime` since that's what the legacy module writes and we share its cache; the docblock now spells that out explicitly. The `get-monitor-status` description also now mentions the legacy projection so the semantics are visible to agents.

**4. Testing instructions.** Replaced the prose steps with copy-pasteable `wp eval` commands for `wp_get_ability(...)->execute([...])`, plus a `curl` for the REST surface. Step 7 explicitly demonstrates the idempotent `changed:true` / `changed:false` pair.

Ready for another look.

**Review by chrisbliss18 — CHANGES_REQUESTED** [review]
Thanks, this is looking much better. The gated-registration story is consistent now, and the negative bootstrap test plus the `set-notifications` success-path coverage close the gaps I was worried about. The `last_status_change` rename and the docblock about sharing the legacy transient are exactly what I was after.

One more thing surfaced while I was thinking about the Jetpack Monitor v2 rollout regarding `last_status_change`:

`fetch_last_status_change` only treats an empty string (or a non-string) as "no transition yet" and passes everything else through verbatim. That's correct against Jetpack Monitor v1 today, which returns an empty string for "none". But when Jetpack Monitor v2 takes over the `last_status_change` projection, I'd like this ability to stay stable no matter how the backend spells "no value". If v2 ever lands a MySQL zero-date (`0000-00-00 00:00:00`) instead of an empty string, the current check would surface that as a literal timestamp string rather than `null`, and the documented `null` contract would break for agents. 

Recommendations:
1. Pull the two inline checks (`class-monitor-abilities.php` lines 333 and 347) into one normalizer and have both branches call it, so "no transition" collapses to `null` regardless of
representation. Something like:

```
private static function normalize_last_status_change( $value ) { 
      if ( ! is_string( $value ) ) { 
              return null;
      }   
      $value = trim( $value );
      if ( '' === $value || 0 === strncmp( $value, '0000-00-00', 10 ) ) { 
              return null;
      }   
      $ts = strtotime( $value );
      if ( false === $ts || $ts <= 0 ) { 
              return null;
      }   
      return $value;
}
```

A legitimate `YYYY-MM-DD HH:mm:ss` value passes through unchanged, so this stays clear of any timezone/format questions. The `strtotime` line is just a backstop so it self-defends against representations we haven't seen yet; if you'd rather keep it strict, the empty + zero-date checks alone are enough.

2. Add a direct test for the normalizer (empty, whitespace, `0000-00-00 00:00:00`, a valid timestamp, and a non-string). `Monitor_Abilities_Test_Stub` currently overrides `fetch_last_status_change` wholesale, so this normalization is the one part of the method with no coverage today. It could be worth closing alongside the change.

Everything else looks good to me. Thanks again for your work on this.

**Comment by enejb** [comment]
Thanks @chrisbliss18 — pushed `6b96226e56` to address both recommendations.

**1. Single normalizer for "no value" representations.** Extracted the two inline `is_string && '' !== $cached` checks in `fetch_last_status_change` into one `normalize_last_status_change( $value )` helper, and both the cache-hit and fresh-fetch branches now route through it. Followed the shape you suggested:

- non-string → `null`
- `trim()` then empty → `null`
- `strncmp( $value, '0000-00-00', 10 ) === 0` (catches `0000-00-00`, `0000-00-00 00:00:00`, and any future zero-date variant) → `null`
- `strtotime` backstop (`false === $ts || $ts <= 0`) → `null` — kept it in as the self-defense layer for representations we haven't seen yet
- otherwise: pass the original `YYYY-MM-DD HH:mm:ss` string through unchanged, so timezone/format semantics stay clear of the helper

So the documented `null` contract is now stable regardless of how Monitor v2 spells "no transition yet."

**2. Direct test for the normalizer.** Added `test_normalize_last_status_change_collapses_no_value_representations_to_null`, a `@dataProvider`-driven test exercising:

- empty string
- whitespace-only (`"  \t\n"`)
- `'0000-00-00 00:00:00'` and `'0000-00-00'`
- `null`, `0`, `false` (non-strings)
- `'not-a-timestamp'` (unparseable)
- `'2024-01-15 12:34:56'` (pass-through)

Since `Monitor_Abilities_Test_Stub` overrides `fetch_last_status_change` wholesale (that's why this path had no coverage before), the test drives the protected helper through a new `expose_normalize_last_status_change()` passthrough on the stub. Keeps the production class's visibility tight and avoids reflection in the test.

Branch is also rebased on current trunk. Ready for another look.

**Review by chrisbliss18 — APPROVED** [review]
Thanks, this addresses the two things I called out in my last review.

The new `normalize_last_status_change()` helper is the shape I was looking for. Both the cached and fresh XML-RPC paths now route through the same normalization path, the no-value representations collapse to `null`, and a legitimate timestamp string passes through unchanged. The data-provider coverage also closes the gap where the test stub was bypassing this path before. I consider my previous `last_status_change` concern resolved.

While doing one more pass, I noticed a few separate follow-ups. These are specific to your PR but separate from what I brought up earlier:

1. Medium priority, probably blocking: `set-notifications` can leave the local `monitor_receive_notifications` option stale. If the remote state already matches the desired value, the method returns `changed=false` before mirroring the known-good remote value locally. That matters because the legacy REST reader trusts the local option before going remote. Since this PR already mirrors the option after a changed write, I think this path should also sync the local option to the remote value it just fetched. The existing `changed=false` test would need to change accordingly.
2. Medium priority, possibly blocking: The abilities set `show_in_rest`, but don't include the `meta.mcp` metadata that the other Jetpack ability definitions I checked include. If these are meant to be visible to MCP/agent tooling, I think this should be added before merge. If they are intentionally REST-only for now, then this is non-blocking, but I'd like that called out explicitly.
3. Low priority, non-blocking: The class lives under `src/abilities` while other module ability implementations are with their module and explicitly required from the module bootstrap file. The current approach appears to work, so I don't think this needs to block the PR, but moving it under the Monitor module would better match the existing pattern.

Everything from my last review is addressed; the items above are separate follow-ups from a fresh pass. My other findings aren't critical, and I don't think they block merging. I did want you to know these details however.

Thanks again for the work on this. :smile: 

I'm signing off as approved, letting you decide if any more work should be done on this PR or just noted for later.

**Comment by enejb** [comment]
Thanks for the thorough review and the approval, @chrisbliss18! I've addressed the two follow-ups from your last pass:

1. **Stale local option on no-op `set-notifications`** (medium / probably blocking): the `changed=false` path now also mirrors the known-good remote value to the `monitor_receive_notifications` option, so the legacy `Jetpack_Core_Json_Api_Endpoints::get_remote_value` reader — which trusts that option before going remote — can't surface a stale state. The unchanged path is now self-healing alongside the existing post-write mirror. I updated the `changed=false` test to seed a disagreeing local option (`false`) against a remote `true` and assert the no-op corrects it.

2. **Missing `meta.mcp`** (medium / possibly blocking): both abilities now carry `'mcp' => array( 'public' => true, 'type' => 'tool' )`, matching the sibling `jetpack/get-modules` and `jetpack/set-module-status` definitions. These are intended to be agent-facing, so MCP-discoverable is the right shape rather than REST-only. Added data-provider test coverage asserting both slugs expose the public-tool metadata.

3. **Class location under `src/abilities`** (low / non-blocking): left as-is for now. The sibling `class-modules-abilities.php` already lives in `src/abilities`, so the current location is at least internally consistent with the other ability registrars. Happy to move it under the Monitor module in a follow-up if we'd rather standardize on the per-module bootstrap pattern.

I also fixed a CI failure introduced by the new test — PHPUnit 11 (PHP 8.2+) requires data-provider methods to be `static` and the `#[DataProvider]` attribute rather than the `@dataProvider` annotation.
